src/jitlayers.cpp

```diff
@@ -736,6 +739,46 @@ class JLDebuginfoPlugin : public ObjectLinkingLayer::Plugin {
         });
     }
 };
+
+class JLMemoryUsagePlugin : public ObjectLinkingLayer::Plugin {
+    private:
+    std::atomic<size_t> &total_size;
+    public:
+
+    JLMemoryUsagePlugin(std::atomic<size_t> &total_size) : total_size(total_size) {}
+
+    Error notifyFailed(orc::MaterializationResponsibility &MR) override {
+        return Error::success();
+    }
+    Error notifyRemovingResources(orc::ResourceKey K) override {
+        return Error::success();
+    }
+    void notifyTransferringResources(orc::ResourceKey DstKey,
+                                    orc::ResourceKey SrcKey) override {}
+
+    void modifyPassConfig(orc::MaterializationResponsibility &,
+                            jitlink::LinkGraph &,
+                            jitlink::PassConfiguration &Config) override {
+        Config.PostAllocationPasses.push_back([this](jitlink::LinkGraph &G) {
+            size_t graph_size = 0;
+            for (auto block : G.blocks()) {
+                graph_size += block->getSize();
+            }
+            this->total_size.fetch_add(graph_size, std::memory_order_relaxed);
+            return Error::success();
+        });
+    }
+};
```

Comment:
Might want to run this through clang-format; the indentation looks off in places.